### ISSUE TITLE
user management with separated auth and resource server 

### ISSUE THREAD

**Issue by ehsmsh** [jazzband/django-oauth-toolkit#1197]
With the separated auth and resource server architecture, username in not a unique permanent identifier to be used in resource server to create users and relate resources to them. It should be user_id which is the primary key of the user, making sure it would never change.

At the same time it should be implemented in introspect endpoint to have user_id in the response, along side the other parts. Right now only username is returned about the user:
```
if token.user:
    data["username"] = token.user.get_username()
```

**Comment by sanxchep** [comment]
Since the oauth table only links to django auth user model, you will have to implement a custom auth table to do that i guess.
But you do have a point, different resource servers may have different set of users with same username, I think due to this, we should have the ability to set different oauth user model to different models that we might create, linked with id ofcourse.

**Comment by ehsmsh** [comment]
You are right about the resource server as it needs to overwrite the basic user model. But for the auth server introspcet url which is used by the resource server to authenticate users, it can be included in the response as default with out a problem. I have overwritten the TokenIntrospectionView as so which i think should be the default toolkit class:

```
@method_decorator(csrf_exempt, name="dispatch")
class IntrospectTokenView(BasicIntrospectTokenView):
    @staticmethod
    def get_token_response(token_value=None):
        try:
            token = (
                get_access_token_model()
                .objects.select_related("user", "application")
                .get(token=token_value)
            )
        except ObjectDoesNotExist:
            return JsonResponse({"active": False}, status=200)
        else:
            if token.is_valid():
                data = {
                    "active": True,
                    "scope": token.scope,
                    "exp": int(calendar.timegm(token.expires.timetuple())),
                }
                if token.application:
                    data["client_id"] = token.application.client_id
                if token.user:
                    data["username"] = token.user.get_username()
                    data["sub"] = token.user.pk
                return JsonResponse(data)
            else:
                return JsonResponse({"active": False}, status=200)
```
The subject of the token is user id in the auth server.

**Comment by sanxchep** [comment]
same tbh, I had to overwrite literally all stages of authentication to make it a "one-auth-server" for everything in my org.
would be able to do it gracefully like django code style would want to... but here we are.

**Comment by ehsmsh** [comment]
Hope they change the introspection view like this, as it is clear. About the validator class it is a little bit out of the toolkit scope because of user model. But the part of the method which creates user after introspection can be separated so the overwrite be more clear and to the point. With the current implementation, the overwrite needs copying a lot of code which is not related to user model creation, bc the method is doing so much all together.

**Comment by ehsmsh** [comment]
to be pricise, this part of _get_token_from_authentication_server method in OAuth2Validator can be separated in a new class method like below. Because this the only part of the validator related to users.

```
def _get_or_create_authenticated_user(self, content=None):
    if "username" in content:
        user, _created = UserModel.objects.get_or_create(
            **{UserModel.USERNAME_FIELD: content["username"]}
        )
    else:
        user = None
        
    return user
```
